### ISSUE TITLE
Feature request: try_cast function

### ISSUE THREAD

**Issue by neverchanje** [risingwavelabs/risingwave#4458]
**Is your feature request related to a problem? Please describe.**

I'm working on a demo that processes JSON data in which there is field called 'value', which can be a number or a string. To identify which type it is and process the two types separately, I will need 'try_cast'.

`create mv mv1 as select avg(value) from t where try_cast(value as int) is not null`

`create mv mv2 as select last(value) from t where try_cast(value as varchar) is not null`

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**
https://dba.stackexchange.com/questions/203934/postgresql-alternative-to-sql-server-s-try-cast-function

**Comment by xiangjinwu** [comment]
related: #4625
However, it will return `null` for the whole expr, rather than the subexpr. Resulting in `cast(value as int) is not null` returning `null` rather than `true`.

**Comment by xxchan** [comment]
> Resulting in `cast(value as int) is not null` returning `null` rather than `true`.

When the cast fails, `try_cast(value as int)` is expected to be `null`, then `cast(value as int) is not null` is `false`. It seems also ok if it's null 😄?

```sql
create table t(x varchar);
insert into t values('1'), ('2'), ('haha');
create materialized view mv as 
select sum(cast(x as int)) as sum from t;

select * from mv;
 sum
-----
   3
(1 row)
```

So this method actually CAN work. But it seems quite hacky (and the performance may be poorer) and I think add a `try` version for fallible functions is more appropriate. 🤔

**Comment by xiangjinwu** [comment]
But `cast(value as int) is null` also returns `null` rather than `true` 😢 Or the user may want to use a non-trivial default value `coalesce(cast(x as int), 42)`.

To list all alternatives that have been mentioned during random offline discussions:
* Implement `try_cast`
* Implement `err_as_null`. This supports `err_as_null('a'::int)` as well as `err_as_null(a + b)` for overflow.
* Update current error tolerance to be per-subexpr. (And possibly exposing a session config.)

**Comment by xxchan** [comment]
`err_as_null` & subexpr error tolerance can be both implemented as wrapping expr with `InfallibleExpression` in `build_from_prost`? 🤔

`err_as_null` is a branch while subexpr error tolerance is at the end.

**Comment by xiangjinwu** [comment]
* `IS JSON` is part of ISO/IEC 9075:2016. PostgreSQL reverted its support before releasing v15 and is adding back in v16. We can reject `WITH UNIQUE KEYS` if not easy.
* `TRY_CAST` is also popular in many databases. It is applicable to more types and avoid parsing twice for validation and consumption.
* `err_as_null` is more general in terms of types / operations, and is finer-grained than existing streaming error handling.

My take is that we can support both `IS JSON` and `TRY_CAST`, and then `err_as_null` when needed.

Also note that `TRY_CAST(expr AS ty)` is not equivalent to `err_as_null(CAST(expr AS ty))`. The former may still return error during evaluation of inner `expr`, while the latter is supposed to catch all errors inside.

**Comment by xiangjinwu** [comment]
The introduction of `CommonSubExprExtractRule` (CSE) has an impact of some of the solutions mentioned:
* Streaming `eval_infallible` may not work as expected because "whole expr" may be broken by optimizer
* `err_as_null` should be treated as a blackbox with no inner exprs extracted

**Comment by st1page** [comment]
The `try_cast` maybe not enough. We found PG's expression can throw more errors than we thought. What we need might be `try()` for these situations.

```
dev=# with t(v) as (values ('{"a":[1,2,3]}'::jsonb), ('{"a":null}')) select jsonb_array_length(v -> 'a') from t;
ERROR:  cannot get array length of a scalar
```

**Comment by fuyufjh** [comment]
As @st1page said, every expression need a similar syntax. But `try()` looks wired, like `try-catch` in programming language but I think it's much beyond the scope of expression. I would rather to propose https://github.com/risingwavelabs/risingwave/issues/11586#issuecomment-1674173083 which is simple and intuitive.

**Comment by xiangjinwu** [comment]
1. `IS JSON` has been supported (except for `WITH UNIQUE KEYS`)
2. `try_cast` may be skipped if we are going to implement the more general solution soon, unless we need it for compatibility.
3. I think the general `try` is similar to `err_as_null` above, or maybe its sugared form `coalesce_err` that rewrites into `coalesce(err_as_null(), ...)`.
4. #11586 means it is transparent in SQL expr rather than explicitly (introducing and) writing new "error-handling" expr.

Common concerns to address for the last 2 solutions:
* interaction with optimizer, for example common sub-expression extract
* handling of errors from expr-like structures: agg / srf / window / subquery

**Comment by xiangjinwu** [comment]
resolved same as #11586
